### DISCUSSION TITLE
EPropCompiler: accessing per-epoch loss and restricting accuracy evaluation to recall window

### DISCUSSION THREAD

**Discussion by Bekkozhan** [genn-team/genn#753] [mlGeNN questions]
Dear James and Thomas,

Following up on our email exchange in March — I've been working through the e-prop implementation in mlGeNN and adapting the evidence accumulation task (Bellec et al. 2020) as a validation benchmark for cross-simulator comparison with NEST. While reading through the source code I had a few questions — apologies if these have obvious answers I missed.

**1. Is per-epoch cross-entropy loss accessible from `EPropCompiler`?**

Looking at `compiled_training_network.py`, `train()` returns `train_metrics` which appears to contain only the accuracy metric. I can see that `SparseCategoricalCrossentropy` is injected into the output neuron model by `EPropCompiler`, but I couldn't find a way to surface per-epoch loss to Python. Is there a way to record it via a callback or otherwise, or is it intentionally not exposed?

**2. What does `get_readout()` return, and over what window?**

In `_train_batch()`, `get_readout()` is called after all `example_timesteps` have run. My understanding is that this reads output neuron state accumulated over the full trial. Is that correct? The evidence accumulation task runs over 2,250 ms, and accuracy in NEST is computed over the recall window only (the final 150 ms). Is there a recommended way to restrict evaluation of output neuron state in mlGeNN to that window, for example via a custom callback?

Also, would you prefer I batch questions periodically, or post them as they arise? I'm happy to work either way.

Thank you for your time.

Bek

**Comment by neworderofjamie** [comment]
Sounds an interesting project! 

> Is there a way to record it via a callback or otherwise, or is it intentionally not exposed?

Not exposing loss to users is an omission in mlGeNN 2. This is definitely not something to do during any type of benchmarking as it's very slow but the solution is to create a ``VarRecorder`` for your output population with ``genn_var="VSoftmaxSum"`` to record the accumulated total of the softmax values or ``genn_var="VSoftmax"`` to record the per-timestep value (this is _reasonably_ well documented at https://ml-genn.readthedocs.io/en/latest/usage/callbacks_recording.html). However, sadly there aren't currently _timestep_ filters on ``VarRecorder`` to let you select only the last timestep of the epoch so if you are looking for "VSoftmaxSum" you will be recording a lot of superfluous information.

>  Is there a recommended way to restrict evaluation of output neuron state in mlGeNN to that window, for example via a custom callback?

This is also something missing from mlGeNN 2 and, honestly, I can't think of a good way of achieving this aside from modifying https://github.com/genn-team/ml_genn/blob/master/ml_genn/ml_genn/compilers/eprop_compiler.py#L388-L391 to do something like:

```python
model_copy.append_sim_code(
    f"""
    if(t > 2100.0) {
        E = {model_copy.output_var_name} - yTrue;
    }
    else {
      E = 0.0;
    }
    """)
```

Obviously, if you wanted to by slightly less hacky, you could control the time range during which errors are calculated via parameters passed to the class and I would be happy to receive a PR implementing that :wink:

**Reply by Bekkozhan** [reply]
Thank you James — this is very helpful. I'll work on adding a parameter to control the time range during which errors are calculated, as you suggested. Will update this thread with a link to the PR once it's ready.

**Reply by Bekkozhan** [reply]
I've implemented the parameter to control the time range during which errors are calculated and tested it with the error starting at timestep 2100 (the recall window only). The network does learn initially but then degrades badly to a chance level. I used the default hyperparameters (eta=5e-3). Do you have any suggestions? I'm wondering if the learning rate needs adjusting when the error window is much shorter than the full trial.

**Reply by neworderofjamie** [reply]
Could you share your code with me? I just dug out the ancient prototype implementation of this task I made and it seems I used eta=5e-3 so I don't think that's the issue

**Reply by Bekkozhan** [reply]
Thank you — here is the implementation: https://github.com/Bekkozhan/ml_genn/blob/eprop-recall-window/ml_genn/ml_genn/compilers/eprop_compiler.py
In my network I've added weight clipping to [-100, 100] after each batch to match NEST's Wmin/Wmax constraint, and I'm using eta=5e-3.

**Reply by Bekkozhan** [reply]
Hi James, 

I hope you had good extended weekend. I've made some progress on the convergence issue and wanted to share the findings:

The root cause turned out to be two separate problems:
1. SumVar readout accumulates over the full trial. With error_start_timestep=2100, the error signal E = VSum - yTrue was computed against a value accumulated over all 2250 timesteps, not just the recall window. This made the gradient signal effectively meaningless. I implemented a WindowedSumVar readout that only accumulates from error_start_timestep onwards: https://github.com/Bekkozhan/ml_genn/blob/eprop-recall-window/ml_genn/ml_genn/readouts/windowed_sum_var.py — happy to open a PR if you think it's a good fit for the main repo.
2. No weight constraint in EPropCompiler. NEST's eprop_synapse enforces Wmin/Wmax=±100 pA after every update. Without an equivalent constraint, Adam accumulates unbounded momentum in recurrent connections, causing periodic network collapse. I implemented a WeightClipCallback (tested  ±100.0; ±1.0 and ±2.0) as a workaround in pynn-learning wrapper:
```python
def on_batch_end(self, batch, metrics):
    for conn in self._connections:
        genn_pop = self._compiled_network.connection_populations[conn]
        genn_pop.vars["g"].pull_from_device()
        np.clip(genn_pop.vars["g"].view, -1.0, 1.0,
                out=genn_pop.vars["g"].view)
        genn_pop.vars["g"].push_to_device()
```
With both fixes the network reaches >90% accuracy within the first 200 epochs — comparable to NEST. However, learning eventually degrades regardless of clip bound. It may be worth considering whether EPropCompiler could expose w_min/w_max as parameters. This may be more efficient than the pull→clip→push workaround in the wrapper.

**Reply by neworderofjamie** [reply]
Sounds like you've been very productive! 

1. This totally makes sense. I think the best solution here would just be to include your new readout with your model - I am fairly sure the eprop compiler doesn't requires any knowledge of readout implementation.
2. Optimisers do actually support clamping of variables (this is used for delay learning with eventprop), although it is not exposed to the user. If you would like to implement this nicely, I would recommend:
    1. Adding a constructor argument to e-prop compiler that takes of dictionary of connection to tuple 
    2. In the constructor 'normalise' the dictionary key with the ``get_underlying_conn`` function and store it in the class
    3. Pass tuple corresponding to connection (or ``None``) to ``_create_optimiser_custom_update`` at https://github.com/genn-team/ml_genn/blob/master/ml_genn/ml_genn/compilers/eprop_compiler.py#L562
    4. In ``get_model``, pass this tuple through to ``Optimiser.get_model`` in place of the None

**Reply by Bekkozhan** [reply]
Hi James, the implementation worked — with error_start_timestep and weight clamping via the optimiser, mlGeNN now converges stably to 91.1% final accuracy on the evidence accumulation task over 900 epochs. Interestingly, the network reaches ~89% from epoch 0 and stabilises quickly, suggesting the combination of clean recall-window error signal and GPU-side weight clamping allows effective learning within the first pass through the data. Thank you for the guidance on the optimiser clamping infrastructure. I've prepared a clean branch with both changes: https://github.com/Bekkozhan/ml_genn/tree/eprop-clamp-pr — happy to open a PR when you think it's ready.

**Reply by neworderofjamie** [reply]
Wonderful! Please do make a PR and I'll review

**Reply by Bekkozhan** [reply]
PR is now open: https://github.com/genn-team/ml_genn/pull/203 — thank you, James, again for the guidance!

**Comment by Bekkozhan** [comment]
One follow-up on Q1: I couldn't find VSoftmaxSum in mlGeNN 2.3.1 — looking at the source it seems the variable would be VSumSoftmax (from SumVar readout adding VSum, then EPropCompiler appending Softmax). Is that correct, or am I looking in the wrong place?"

**Reply by neworderofjamie** [reply]
Apologies, it is indeed ``VSumSoftmax``
